### DISCUSSION TITLE
Cannot Resolve component specified in MDXProvider

### DISCUSSION THREAD

**Discussion by gaurishhs** [mdx-js/mdx#2225] [Q&A] [answered]
Hello! I'm developing a SSG using Vite and MDX. I'm passing some components in `MDXProvider` but it always gives me the error that component is not passed via props. 

Here is the Provider: https://github.com/gaurishhs/documentum/blob/main/packages/documentum/src/theme/DocsLayout.tsx#L45
Bundler: https://github.com/gaurishhs/documentum/blob/main/packages/documentum/src/node/plugin-mdx/pluginMdxRollup.ts#L33 

Please help me out..

**Comment by wooorm** [comment, marked as answer]
hi!

- You probably don’t need a provider. Pass the components directly.
- You use `jsxImportSource` for `@mdx-js/react`. That is wrong. `jsxImportSource` is something different. A provider is not a JSX import source. If you use react, you don’t touch `jsxImportSource`. If you want a provider, you use `providerImportSource`: https://mdxjs.com/packages/mdx/#optionsproviderimportsource

**Reply by gaurishhs** [reply]
How do i pass the components directly? Moreover whenever i remove `jsxImportSource` i keep getting a bunch of other errors

**Reply by wooorm** [reply]
Then provide more info. What errors you're getting.

How to use mdx is in the docs on "Using mdx": https://mdxjs.com/docs/using-mdx/

**Reply by gaurishhs** [reply]
Just tried. It works without specifying `jsxImportSource`. Please guide me on passing the components directly.

**Reply by wooorm** [reply]
https://mdxjs.com/docs/using-mdx/#components

**Reply by gaurishhs** [reply]
Well, It doesn't seem to work.. 

```jsx
import {
  DocsContent,
  DocsHeader,
  DocsSidebar,
  TabItem,
  Tabs,
  Admonition,
  Code,
  CodeGroup
} from 'ui/lib/esm';
import { MDXProvider } from '@mdx-js/react';
import { Fragment } from 'react';
import { Outlet } from 'react-router-dom';
import { useAppContext, useSidebar } from '@theme/hooks';

export function DocsLayout() {
  const {
    title,
    themeConfig: { navLinks, logo, footer, sidebar }
  } = useAppContext();
  const { isOpen, setSidebarState } = useSidebar();
  const toggleSidebar = (s?: boolean) => {
    if (s) {
      setSidebarState(s);
    } else {
      setSidebarState(!isOpen);
    }
  };
  return (
    <Fragment>
      <DocsHeader toggleSidebar={toggleSidebar} navLinks={navLinks} />
      <DocsSidebar
        title={title}
        sidebar={sidebar}
        setState={toggleSidebar}
        isOpen={isOpen}
        navLinks={navLinks}
        logo={{
          dark: logo?.dark,
          light: logo?.light
        }}
      />
      <div className="dark:bg-zinc-900 bg-white">
        <DocsContent
          components={{
            Tabs,
            TabItem,
            Admonition,
            Code,
            CodeGroup
          }}
          footer={footer}
        >
          <Outlet />
        </DocsContent>
      </div>
    </Fragment>
  );
}

```

Tried passing the components to `DocsContent` and to `Outlet` as well but it doesn't work..

**Reply by wooorm** [reply]
Are they mdx files?

If you send post to one friend, do you expect it to be delivered to your other friend?

You can pass components to an mdx component. Passing then to random other components doesn't do anything.

**Comment by gaurishhs** [comment]
No they are not.. Im not importing mdx files they get compiled to Jsx ones

**Reply by wooorm** [reply]
Yes, they are compiled from a string of MDX to a component. You can pass components to those components.

**Reply by gaurishhs** [reply]
Can't i use a Provider, Instead of directly passing

**Reply by gaurishhs** [reply]
This is what i did. But I'd really prefer a `Provider` based solution

```jsx
import { useRoutes } from 'react-router-dom';
import { routes } from 'dcm:routes';
import { DocsLayout } from '../theme/DocsLayout';
import NotFound from '@theme/NotFound';
import { TabItem, Tabs, Code, CodeGroup } from 'ui';

const components: Record<
  string,
  (({ children, ...props }: { children: any }) => JSX.Element) | typeof Tabs
> = {
  TabItem,
  Tabs,
  Code,
  CodeGroup
};

const updatedRoutes = [
  {
    path: '/',
    element: <DocsLayout />,
    children: routes.map((route) => {
      const MDXComponent = ({ components }: any) => {
        return route.element;
      };
      return {
        path: route.path,
        element: <MDXComponent components={components} />
      };
    })
  },
  {
    path: '*',
    element: <NotFound />
  }
];

export const Content = () => {
  const routeElement = useRoutes(updatedRoutes);
  return routeElement;
};
```

**Reply by wooorm** [reply]
Does that work? It doesn’t look like it works. You define a component yourself, called `MDXComponent`, but it isn’t an `MDXComponent`? And it just drops all the props that are given?
What is `DocsLayout`? Where do you import MDX files?

Why do you want to write more lines of code in your own app, that includes more code in the client side, which is also a lot slower? And it won’t work in React Server Components.

There is one use case for providers: if you import MDX inside other MDX files and find yourself passing components around a lot. I don’t see you doing that.

**Reply by gaurishhs** [reply]
I do not import the MDX files they are compiled by `@mdx-js/rollup` on the server side then the js produced is imported here https://github.com/gaurishhs/documentum/blob/366ab1ee54a4f60c2141f18c446f444bdef75791/packages/documentum/src/node/plugin-routes/RouteService.ts#L65 and then using `React.createElement` a element is created [here](https://github.com/gaurishhs/documentum/blob/main/packages/documentum/src/node/plugin-routes/RouteService.ts#L74) which is further used in the client-side. Should i be passing it in props in `createElement`?

**Reply by gaurishhs** [reply]
`DocsLayout` is basically a layout for every element

**Reply by wooorm** [reply]
Can you please please show how you import (the compiled) MDX?

**Reply by gaurishhs** [reply]
```ts
3  |  import loadable from "@loadable/component";
4  |  
5  |  const Route0 = loadable(() => import('/Users/gaurish/Desktop/Coding/documentum/docs/getting-started/configuration.md'));
6  |  const Route1 = loadable(() => import('/Users/gaurish/Desktop/Coding/documentum/docs/getting-started/installation.mdx'));
7  |  export const routes = [
```

**Reply by gaurishhs** [reply]
Then in routes array `React.createElement` is used to create elements out of those loadable imports

**Reply by gaurishhs** [reply]
> There is one use case for providers: if you import MDX inside other MDX files and find yourself passing components around a lot. I don’t see you doing that.

Is this necessary? I find implementing direct passing of components much more difficult than simply using a provider. Moreover i wouldn't prefer importing JSX Elements on client-side..

**Reply by wooorm** [reply]
Can you show the `React.createElement` call?
You can pass the components there. Because `Route0` and `Route1` are the MDX components. They accept components.

> Is this necessary

Using fast and less code is not necessary, but it’s a good idea.

> much more difficult than simply using a provider

It is not more difficult. It is simpler. There is less code.

> Moreover i wouldn't prefer importing JSX Elements on client-side..

We probably have different ideas about what “JSX elements” are. You don’t need to use anything client side with MDX. And if you do, you already are.
JSX is compiled away. Just like MDX. No JSX ends up on the client.
If you mean React nodes, then there is no way to remove them on the client (other than SSG / SSR, which does not affect this setup, or RSC, which can’t be used with providers)

**Reply by gaurishhs** [reply]
Alright. Thank you so much for sparing your time. I passed the components to the `React.createElement` call and it works.

**Reply by wooorm** [reply]
No worries! Good luck on your endeavours! :)
